osx_defaults: refactor to use Foundation/CoreFoudation python modules instead of `defaults` command, adding a lot of new features at the same time (nested keys and values, support for all types, etc.)

etienned opened the pull request:
##### SUMMARY
Update of the osx_defaults module to use Foundation / CoreFoudation python modules instead of the `defaults` command. The goal is to add new features and those features were not possible using the `defaults` command. New features are:
- Access to nested keys and indexes (ex.:  DesktopViewSettings -> IconViewSettings -> arrangeBy).
- Writing nested and complex values (ex.: [1, {key1: 'text', key2: true}, 'other stuff']).
- Support for all data types: array, boolean, data, date, dict, float, integer, string.
- Support to access preferences for any users (like network settings, etc.).
- Specifying type is now optional. Types are mostly deduce as YAML/Ansible do.

Those changes should be backwards compatible.

Also this update fixes some issues that are present in the current implementation (`array_add` not keeping items order, etc.).

Code cleanup to adhere to Ansible's guidelines.

Not python 3 compatible for the moment. But in most case Foundation / CoreFoundation modules runs under the stock macOS' python and it's currently version 2.7. It should not be too hard to add support for python 3 after the module will be tested for some time in the wild.

The new version of the module is built in way that it could be relatively easy to add other backends like `defaults` and `PlistBuddy` commands. It could be possible to automatically fallback to another backend (ex. using `defaults`, so module will have with limited functionalities) if Foundation can't be imported. I don't know if it's a good or a bad idea.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
osx_defaults

##### ANSIBLE VERSION
```
ansible 2.4.0 (devel da99deeddc) last updated 2017/05/18 22:45:57 (GMT -400)
```


##### ADDITIONAL INFORMATION
I know there's already at least one other pull request (like [this one](https://github.com/ansible/ansible/pull/24032)) for an updated version of `osx_defaults` but I think this version is more complete. I started to work on this a long time ago after [I submit](https://github.com/ansible/ansible-modules-extras/pull/1444) a fix for a bug. Unfortunately I didn't have a lot of time to work on it. I also did many iterations, first one was using `PlistBuddy`.

I did a mostly complete test suite for the main class (`CFPreferences`) but unfortunately at the time I start it there was no official way to test modules. Also it was not possible to import something, like a class, from an Ansible module. So my test suite is built on `pytest` and the class is simply copied in the test file. That's not really pretty. You can find it [here](https://gist.github.com/etienned/34d8247756a1a49110c08385b85d634b). If there's a relatively simple way to integrate this test suite to Ansible, tell me. I also have a test playbook that cover most combinations [here](https://gist.github.com/etienned/ced131cbf8c34ddf3e1a09d6a945812f#file-test_osx_defaults-yml).

I would like also to mention that I had to circumvent some inconsistent Ansible behaviors, particularly on how Ansible cast types in YAML, as this module use YAML structure and types intensively. I know there's some work that is done on that side currently and it's a good thing. I hit some limitations of JSON: no support for datetime and for data. That's one of the reason that datetime should always be quoted for this module. Maybe a kind of support for those types could be added directly to Ansible in the future.

I tried to document the code correctly but if you have any interrogations just ask.

etienned committed:
osx_defaults: refactor to use Foundation/CoreFoudation python modules instead of `defaults` command, adding a lot of new features at the same time (nested keys and values, support for all types, etc.) and python 3 support.

booch commented:
This rewrite has everything I've dreamed of for the `osx_defaults` module, and more. There are a lot of situations where I still need to fall back to using the `defaults` command. This rewrite resolves them all. Very nicely done! I'd really like to see this go into 2.5 ASAP.

booch commented:
So what's the process for getting this included in the Ansible 2.5 release? What can we do to help the process along?

gundalow commented:
As agreed in 2017-12-07 Core meeting, we've done bulk pep8 fixes across all modules in `devel`.

This was done in https://github.com/ansible/ansible/pull/33682 and https://github.com/ansible/ansible/pull/33690 which have been merged into `devel`. because of this you will need to rebase your PR. Thanks in advance.

etienned commented:
I've squashed and rebased my PR, as asked.

kinglouie commented:
I have a similar module PR open and just found your refactor which looks a lot more sophisticated than my PR https://github.com/ansible/ansible/pull/35683

Just one feature I like about my version is the ability to deep merge dicts, maybe you could add that? 
I would also suggest to put this into a new module since the name `osx_defaults`suggests that the defaults command is used, also "OS X" was renamed to "macOS", what do you think about this?

kinglouie commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
-    required: false
     default: null
     version_added: "2.1"
+  any_user:
```
I would suggest renaming this to `user` with choices `["currentUser", "anyUser"]` and `default: currentUser` for better consistency with host.

etienned replied:
As I don't think I'll make your previous suggestion about `host`, consistency is not an issue here. Also, as there's only two values possible, it's simpler for the users to treat it as a boolean, so users don't have to remember what are the choices.

kinglouie commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
+      - The I(domain) is a domain name of the form com.companyname.appname or
+        I(NSGlobalDomain) for the Global Domain.
     default: NSGlobalDomain
   host:
```
I suggest adding `choices: ["anyHost", "currentHost"]` with  `default: anyHost`

etienned replied:
I don't think I should do this for two reasons:

1. I want the `osx_defaults` module to be backward compatible, so peoples already using it do not have to rewrite their tasks to use this module that have more features.
2. It's possible to specify a specific hostname as host, so using `choices` is not an option.

kinglouie commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
+from ansible.module_utils.six import string_types, binary_type, text_type
+from ansible.module_utils.six import integer_types
+
+try:
```
To ensure pyObjC lib is available even if the user has a homebrewed python version maybe add something like this: ```sys.path.insert(0, '/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/PyObjc')```

etienned replied:
By default, Ansible use the python located at `/usr/bin/python`, so, in macOS, this python is the system provided python that already have PyObjC. No problem in this case.

If someone specify in Ansible another python to use (like a homebrewed python3) I think it's his responsibility to insure that this python have PyObjC installed (just adding the system PyObjC to path will not work with another python version anyway).

Does this make sense?

azaghal replied:
I think it makes sense - the user is responsible for making sure the library is available for specified Python interpreter. Similar to how user is required to install necessary prerequisites for modules, for example.

etienned commented:
@kinglouie First, thanks for taking the time to review my PR! It's appreciated.

Now about your `deep merge` suggestion: I think my module already do something similar but it's not exactly the same. With my solution you can specify a nested key as this:
`key: 'DesktopViewSettings:IconViewSettings:arrangeBy'`.

So it's possible to change only what you want without rewriting all the structure. The difference with your solution is that, in my case, it's not possible to change (merge) many dictionary values in one shot. You need to do it one after another. But for an array, it's already possible to do it in one shot by using `array_add`. So I will wait to check if there's really a need before adding this feature because it will add complexity to a code that's already enough complex.

Concerning the name change, my idea with this module was definitely to be backward compatible with the existing module for different reasons:

1. Users will not have to rewrite their playbooks.
2. Module is already known, so no need to discover a new module.
3. Two modules that mostly duplicate the same functionalities sound a bad idea for the users: which one to choose, etc.
4. One module less mean less maintenance for the Ansible community.

Your idea of returning the read values are definitely interesting. But as the value can be a complex data structure with different data types (and JSON do not support binary and date types), this need to be thoroughly thought before adding it.

You can check your other comments, I already have answered them.

azaghal commented:
Hm... I would kind of like to support the idea of joining forces together if possible :) Haven't actually thought about looking into whether there already is a module for setting preferences.

etienned commented:
I'm definitely not agains't the deep merge functionality but, honestly, currently I don't want to invest more time in this module before I'll will know it will be integrated in Ansible. I've done it in my spare time (something that I mostly don't have!) and submitted it almost 9 months ago and I didn't really received real feedback from Ansible's committers in all this time. So I'll wait for real feedback before working on this.

I should add that your python code is not bad at all for someone who started two weeks ago!

kinglouie commented:
I totally understand that you are a frustrated that your effort went unnoticed by the maintainers. In fact as I was researching if there were any PRs on this matter, I found a few other promising ones that all got stale because they were lacking response from maintainers. I think to get a little tranction here, we have to advertive it on the IRC meetings waiting alone won't help.

Back to your implementation: 
Is it possible to change the type of a key? On a fresh macOS in the `currentHost com.apple.screensaver` domain, the type of the `idleTime` key is `string`, if you change this setting in System preferences it gets converted to the type `number`.
Using your implementation to change the idletime I get a type mismatch error. Maybe we should add some kind of option to `replace` current key or to `force` values.

etienned commented:
About advertising on IRC, this could be a good idea but currently I don't have the time to do the follow up if it get attention.

About my implementation, normally it should raise an error if there's a type mismatch, as the defaults command and the current osx_defaults module do also (so my implementation is backward compatible).

But I should say that I find pretty strange that a key changes is type. In general, software that read those preferences expects to always have the same type for a specific key. I took a look at my system and the `com.apple.screensaver` `idleTime` under `currentHost` was not defined. After I change this setting in System preferences it was added as an `integer` (not `number` because there's two types of number: `integer` and `real`). So my guess is that your system was not totally fresh and that you had set `com.apple.screensaver` by yourself with the wrong type without noticing. But I could be wrong, of course. Out of curiosity, how did you read pfiles?

I should add that I've use my module to change a lot of preferences and every time I had a type mismatch error, it was my fault.

etienned commented:
ready_for_review

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
+    - OS X caches preferences aggressively. This module should take care of
+      updating caches but in some cases you may need to logout and login to
+      apply the changes.
+    - Check mode can be use with this module.
```
Typo - `use` -> `used`. (your friendly neighbour nitpicker :)

etienned replied:
I'm happy to have a nitpicker! English is not my first language and I'm always happy to improve it.

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
+      apply the changes.
+    - Check mode can be use with this module.
+    - Nested keys need to be quoted when use in abbreviated form
+      (because they contains colons and that mess with the syntax).
```
Again nitpicker - sentence sounds a bit weird. Maybe something as follows would be better?

```
Nested keys must be quoted when specified in abbreviated form (e.g. C("DesktopViewSettings:IconViewSettings:arrangeBy")), since colons are treated as part of YAML syntax itself.
```

etienned replied:
Definitely better!

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
+
+# Regular expression that match datetime formats. Should match mostly the
+# same formats that YAML support.
+RE_DATETIME = re.compile(r"""
```
I wonder if this whole regex thing for datetime would be better off done away in favour of standardising the date format and just using `datetime.datetime.strptime`?

It could be maybe possible to support a couple of different formats and determine which one to use based on target string length instead. That is - multiple possible `strptime` formats, each longer than the previous one (so, one for `YYYY-MM-DD`, one for `YYYY-MM-DD HH:MM` etc.

But again, I would think one format might be better off. Another thing is that if you specify partial date, your actual UTC date/time might be changing based on which timezone you run the code in, right?

etienned replied:
The main reason why I've used a regex is because it's complicated to [support the timezone with `strptime` in python 2.7](https://stackoverflow.com/a/38992623/146481) (this is just an example, we also need to supports 'Z', etc.).

It gives also a bit of flexibility on the format. I should mention that this regex is adapted from the one defined in the [YAML specifications](http://yaml.org/type/timestamp.html), so it should be pretty safe and it's not really complex. It is also virtually conforming to the [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) format, a format that gives this flexibility.

About the partial date, if you mean a date without a timezone (a naive date), my code add the local timezone when there's no timezone. This mimics how the `defaults` command works.

azaghal replied:
Fair enough, I just tend to be careful about complex regex'es. Maybe then link to their page too, might be useful for someone else who reads to know why the regex was picked?

etienned replied:
Yes, I should add the references.

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
+            int(match.group('hour')) if match.group('hour') else 0,
+            int(match.group('minute')) if match.group('minute') else 0,
+            int(match.group('second')) if match.group('second') else 0,
+            # Skip microsecond because they are unsupported in CF.
```
Should this be mentioned somewhere in documentation?

etienned replied:
Yes, it's a good idea.

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
-    # tools --------------------------------------------------------------- {{{
-    """ Converts value to given type """
+    # List of all text characters.
+    TEXT_CHARS = bytearray(set([7, 8, 9, 10, 12, 13, 27]) | set(range(0x20, 0x100)) - set([0x7f]))
```
What is this exactly supposed to be? ASCII chars? Something else?

etienned replied:
It's bytes used to define text chars present in most encodings (ASCII, UTF-8, ISO-Latin-1, etc.) and a few control characters that are normally present in text as the new line char, etc. All other control characters (null, start of heading, etc.) are absent of this list because they are never used to defined text chars.

This constant is use to find the proportion of control characters vs all characters in the data.

azaghal replied:
I mentioned an idea below (well, more like restated your own) of introducing a prefix for base64-encoded data, so user can easily mark those as such. So maybe that'd sort this part out too.

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
 
-        raise OSXDefaultsException('Type is not supported: {0}'.format(type))
+        # Check if data is a valid base64 string. Short strings are not
+        # considered as binary.
```
Why are short strings not considered to be binary data? Why is value `51` used as a border for this purpose?

etienned replied:
Because with a small dataset, it's hard to know with a good certitude if it's binary data or not (check next comment for the "fox" example). The more data you have, the more chances you have to hit non-text characters.

The 52 value is a bit arbitrary but I found that you rarely have useful binary data that are less then 37 bytes long (52 bytes b64 data converted to binary will gives 37 bytes). For example a 8 by 8 pixels 8 bit images will take 64 bytes. I did some tests on real data in the context of the macOS preferences to find this "52". Shorter than that gives a lot of normal strings detected as binary. I wish there was a better solution but I didn't found one.

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
+        # considered as binary.
+        if isinstance(data, binary_type) and len(data) > 51 and not data.translate(None, cls.BASE64_CHARS):
+            try:
+                binary_data = b64decode(data)
```
I don't quite understand why this is not the default thing attempted to identify base64-encoded binary data?

etienned replied:
Because some really normal character strings are valid base64-encoded binary data. Ex.:

`>>> b64decode('The quick brown fox jumps over the lazy dogs')
b"N\x17\xaa\xba'$n\xba0\x9d\xfa1\x8e\xe9\xa9\xb2\x8b\xde\xae\xd8^\x95\xac\xf2v\x88,"
`

azaghal commented on lib/ansible/modules/system/osx_defaults.py:
```diff
@@ -15,413 +16,975 @@
-        elif self.host == 'currentHost':
-            return ['-currentHost']
+    @classmethod
+    def is_binary(cls, data):
```
This looks like a rather provisional check to me. Why just the first 512 characters/bytes? Why 30%?

etienned replied:
This part is completely inspired by this:
http://code.activestate.com/recipes/173220-test-if-a-file-or-string-is-text-or-binary/

The 512 chars are just an optimization because there's very little chance that binary data will not have a good proportion of non-text characters in the first 512 characters.

For the 30%, I just follow the reference. I can't justify this number.

Here also, if someone have a better/simpler solution I'm all open. I did a lot of researches on this and it's the best solution I found.

azaghal replied:
Any chance on simply requiring that anywhere where some kind of binary data *might* be provided that *all* such data must be `base64`-encoded? Even if it's actual text? Can you identify if property can/should be binary data?

etienned replied:
My first motivation to work on this module was to add access to nested values and to be able to support dicts and lists. Lists and dicts can be nested. Supporting lists and dicts imply two things:
- It's no longer possible to easily specify the type of each items contains in dicts and lists.
- You need to support in some ways all the types supported by the plist because you need to compare existing lists/dicts to the wanted values and those lists/dicts can contains values of any types. That means supporting binary data and datetime.

My first idea to answer the first point was to use the YAML casting property to send values already in their own type. Unfortunately, it just partly works because, for example, values are sent to the host in JSON and JSON do not support binary data and datetime. So I'm sending them as string and deduce their type on arrival.

A nicer solution to this problem would be that Ansible will define a way to send binary data and datetime in JSON (for example by using a marker at the start of the string that tell the type) or by sending YAML instead of JSON.

To answer your question more directly: it's already required that all binary data should be `base64`-encoded. I can identify if a property should be binary only if it's already existing in the plist. For new values, there's no way to know. plist files are just dict/list for storing values. Each app treat plist as they want.

azaghal replied:
Hm... How about then introducing a prefix for such settings at least for this module? So, if user wants to provide binary data to be stored in property, they'd have to say, e.g. 'B64:blah' (where blah would be base64-encoded data). Caveat being that if users wants to store string which starts with B64:, they'd have to do something along the lines of B64:QjY0Og== (QjY0Og== being base64-encoded B64:, I think you get the picture :)

Now, truth be told, not sure if this would be more acceptable for Ansible as project, but I'd see it as more consistent. Maybe second opinions on this could be useful too :)

etienned replied:
That's an idea. Another option for the prefix could be `!!binary`. It's [what is use in YAML](http://yaml.org/type/binary.html). So, if one day Ansible support sending binary data, people will just have to move the first quote after the `!!binary` (`"!!binary R0lGODlhDAA"` -> `!!binary "R0lGODlhDAA"`). Definitely not sure if it's a good idea!

But second opinions is definitely needed on this.

azaghal reviewed (commented):
Well, I tried going through entire PR, but gave up mid-way.

There are huge parts of the code that I find very weird, and it almost feels like the whole thing should be rewritten a bit to reduce complexity, and avoid some of the defensive coding. There's data mangled in unusual way, assumptions made, exceptions caught and ignored all over the place.

It could be also because you are trying to add a huge hunk of new functionality while doing a rewrite on how the changes are made to configuration. Perhaps it would be better to rewrite the existing functionality as separate PR, and then try to introduce new functionality in a new PR?

Please don't get discouraged by this comment, because it can also be the problem is in me, I am genuinely trying to be constructive :)

Could be useful if someone else could chime in on the code presented.

etienned commented:
@azaghal I'm really glad you did review my code! And sorry for the answering delay.

Here I'll try to answer your general comment.

First, this PR is a mostly complete rewrite of the module (but keeping it backward compatible) so I think it's best to review the results instead of the changes. It's a big rewrite because the original version was a simple wrapper around the OS X `defaults` command and to add new features, it was necessary to change this to using the Foundation/CoreFoudation modules directly (you can check in the Issues and PR, there was other propositions in this sense).

Secondly, the large part of the complexity come from the fact that this module is mostly saving values with defined types and to support that, types need to be transfer through all those layers and circumventing their limitations and problems: YAML -> Ansible -> JSON -> Python -> PyObjC. A lot of the code is just for that.

Another part of the complexity is that my version support two more types (binary and dates) than the original version. Those two types are definitely more complex to support then the others (bool, int, float, string). For example, dates and binary are not supported by JSON (used by Ansible to transfer data from the host to the server), so I needed to find solutions for that.

And to be honest, in my version, I decided that type specification was optional (when not specified, YAML types are used, except when problems of Ansible/JSON need to be circumvent) and this definitely contribute to the complexity. For my defence, when I started implementing this I wasn't aware of all the roadblocks Ansible would add me.

Concerning ignore Exceptions, they are use only in autocasting functions where some tries need to be done to deduce the type, so they're not all over the place ;-) I think it's OK to use them in those circumstences. But I'm totally open to better solutions.

It could be a good idea to do this rewrite more incrementally. But I'll need a more tight feedback loop that I had until now (maybe I'm responsible for this lack of feedback, I don't know).

I should also mention that there's currently this PR #35683 that is currently review. This PR is mostly my PR with some propose changes that I did not want to do for two reasons:
 - to not increase complexity at this time
 - to keep to module backward compatible with the original.

I'll try to review your other comments along the week (not too much free time currently, sorry for that).

Sorry for the long answer and thanks again for your time.

azaghal commented:
No need for apologies, it's a community effort, and that's what matters in the end :)

I'll maybe try to provide some more timely feedback from my side (no promises :), but it'd be great if someone with a Mac could actually also go through this and test it. As a general note, perhaps in a couple of places it would be better to be a bit more strict with things that are acceptable input, just to avoid some of the complexity. I think most of the concrete comments I provided are applicable so far. Glad to see you're looking into this, and thank you for trying to improve Ansible - although I do not use Apple stuff myself :)

etienned commented:
Failed tests are not related to my pull request. I've rebased my PR and added some changes suggested by @azaghal.

etienned commented:
ready_for_review

mattclay commented on test/sanity/validate-modules/ignore.txt:
```diff
@@ -786,8 +786,6 @@ lib/ansible/modules/network/ios/ios_logging.py E324
 lib/ansible/modules/network/ios/ios_vlan.py E325
 lib/ansible/modules/network/ios/ios_vrf.py E325
-lib/ansible/modules/network/iosxr/iosxr_command.py E322
-lib/ansible/modules/network/iosxr/iosxr_logging.py E324
```
Restore these lines since these files were not modified in this PR.

etienned replied:
Ha, what a stupid mistake I made. Sorry about that.

etienned commented:
@gundalow Do you think we could ask someone directly if it's interested in reviewing this PR? I ask this because I found this [person](https://github.com/geerlingguy) that work a lot with Ansible, use the `osx_defaults` module and have already contribute to Ansible but I don't know him.

gundalow commented:
@geerlingguy hi, I believe you've used `osx_defaults` a bit, would you be able to look at this PR?

elliotweiser commented:
I can look into testing this when I get some free time. This looks like an interesting change with a lot of potential for flexibility in the future.

I'm a little concerned about this though:

https://github.com/ansible/ansible/pull/24808#issuecomment-355845008

As of Ansible 2.5, Python 3 support was made official, and so it's reasonable for users to expect that promise going forward. The argument that Python 3 support is unnecessary on a Mac _because it ships with Python 2.7_, lacks perspective. AFAIK, most Linux distros still ship with Python 2.7 by default and most (if not all) Linux-based Ansible modules are compatible with Python 3. 

In my experience, Mac users often have to install newer versions of Python (sometimes 2 and sometimes 3) in order to effectively use Ansible. In some cases, I've seen Ansible as a dependency for projects that advertise MacOS-compatibility, but where Python 3 is _required_. In short, I'd hesitate to green-light this if it forces users to go back to Python 2.7.

Aside from that ☝️, I'd be happy to BETA this and/or help make this compatible with Python 3 to help push this past the finish line!

etienned commented:
@elliotweiser I'm glad that you can look at my PR!

About your concern ([#24808 comment](https://github.com/ansible/ansible/pull/24808#issuecomment-355845008)), I'll try to elaborate more on this.

**TLDR**: my version of this module is already ported to python 3 and working but not well tested under python 3 but, for the moment, this module have really low chance to be used under python 3.

First it's important to understand that the code of this module will run only on remote Macs and not on the machine running Ansible. So currently it's already possible to run Ansible with python 3  and using my version of this module without problem.

Secondly, by default, the `osx_defaults` module on the remote will use the python located at `/usr/bin/python` (as most Ansible python modules). For macOS this means the system python (currently version 2.7.*). It's possible to set the `ansible_python_interpreter` settings to another python version. So someone could want to set this to a python 3 version. But my module version have as requirements the `Foundation` and `CoreFoudation` python modules (from [PyObjC](https://pythonhosted.org/pyobjc/)). This serve as a bridge between Python and Objective-C to access macOS functionalities. It's good to note that PyObjC is pre-installed by Apple. So in most cases, the simplest thing to do is to use the system default python that already have those PyObjC modules.

But if someone, for a reason or another, want to use a python 3 version on the remote, it's possible to do so. But the Foudation and CoreFoundation will have to be installed first (they can be installed with a simple `pip install PyObjC`). So here my module will also work because I already ported it to python 3 (I had no choice if I want to passed all Ansible checks). It's just that it's not well tested currently. That's the only thing left to do.

I wish this is more clear now.

elliotweiser commented:
Excellent! Better than I realized! 👍 

Thank you for the clarification. The distinction between the remote vs. local Python installations is important and I'm glad that you made that explicit. I would make one correction to what you said, though. I agree that is _typical_ for the Ansible control node to manage remote machines, but that is not always the case (i.e. `ansible_connection=local`). In my world, this is actually quite common.

Disclaimer: I'm probably just hyper-sensitive about this because I've been bitten by it before. But as you said, if the underlying code is already Py3-compatible and all that's missing to use a customer interpreter is to install the missing package, then I'm sure it'll be fine.

I'll check back when I've had some time to experiment.

etienned commented:
As for using `ansible_connection=local`, it's how I use my module most of the time. In this case, Ansible, when running a module, is treating the local computer mostly the same way as a remote. So, if you have not configured `ansible_python_interpreter`, the module will run under the python located at `/usr/bin/`, even if you're running Ansible with python 3. And I think this is how this module should be use in general.

Anyway, I think I'll have some time in the next weeks to do complete tests of the module under python 3.

etienned commented:
I completely tested the module running under python 3 and found some bugs with the Data class and one bug when converting a python date to a Core Foundation one. All those bugs are now fixed so the module should be python 3 ready. I updated my [test suite](https://gist.github.com/etienned/34d8247756a1a49110c08385b85d634b) for python 3.

mattclay commented:
@etienned Could you include integration tests for this module? Ideally written as a typical Ansible role based integration test -- but you could also use a script based test to run your pytest test suite after making sure pytest was installed.

I'd suggest adding the tests as unit tests, but those need to be cross platform -- so unless you mock all of the macOS dependencies you're better off with an integration test instead.

etienned commented:
@mattclay Sorry for the delay. I think I could include tests. For info, I did not included my tests before because at the time I started to work on this module, in 2016, Ansible did not have tests infrastructure for modules.

What is not clear to me is how to setup the tests so they will be run exclusively on macOS hosts? Everything in this module is linked to macOS preferences system so tests should definitely be run only on macOS.

kinglouie commented:
you can look at my PR, there is a test setup which only runs on macOS.

gundalow commented:
bot_status

etienned commented:
@kinglouie I've checked your tests in your PR and it's pretty clear. Thanks!

dagwieers commented:
@etienned Can you check which of the existing open osx_defaults issues your refactor would solve ? We could ask those reporters to test if you branch fixes their issue as well.

https://github.com/ansible/ansible/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+osx_defaults

jamesdh commented:
Just bumped into a need for this today. Please don't let this die like #24028 did :(

etienned commented:
Unfortunately, I'm pretty busy right now but I'm not forgetting it!

kinglouie commented:
I would be really glad if this could get the attention it deserves. I'm really grateful for the effort you put into this @etienned but I think with the limited time you can offer to support this PR it would be beneficial for everyone who wants to use this, if you could approve that someone else finishes this. It's already 1,5 years later after I pushed an adaption of this PR because at the time it was already stale in here.

dsedivec commented:
This looks like a big improvement upon the existing osx_defaults module, thanks!

One possible problem: Consider domain `com.apple.symbolichotkeys`, key `AppleSymbolicHotKeys`, which is a dict with ints as keys:

```plist
{
    AppleSymbolicHotKeys =     {
        32 =         {
            enabled = 1;
            value =             {
                parameters =                 (
                    65535,
                    126,
                    262144
                );
                type = standard;
            };
        };
};
```

Trying to one of the keys of `AppleSymbolicHotKeys` with something like `key="AppleSymbolicHotKeys:32"` results in an error like:

```
Type mismatch between the key `32` and the node `{...long value...}` (<type 'list'> -> <type 'dict'>).
```

I believe this module thinks the integer key means it should expect a list, which is not true in this case.

Akasurde commented:
Hi @etienned, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.general<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
